### PR TITLE
[docs] Add section on outcome limits

### PR DESCRIPTION
**Pull request by geoknee**
Mostly getting the recent experimental findings and new constants documented in one place. Welcome comments on the content / decisions. 

Also this is located quite early on in the "protocol tutorial" which is probably not the best place, as the concepts are reasonably advanced and belong on material not yet covered at that stage in the tute. 

[Andrew] ... let's see if this
Fixes statechannels/project-management#3

### CODE REVIEW

File: packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md [hunk 1]
```diff
@@ -107,9 +107,23 @@ As a state channel participant, it is advised to check the `FixedPart` of any ch
...
+
+Paying out tokens from a state channel is potentially one of the most expensive operations from a gas perspective (if the recipient does not have any already, the transaction will consume 20K gas per pay out). The same is true of channels paying out (ETH or tokens) to other channels on chain. Bear this in mind when deciding whether to transfer one, many-at-a-time or all-at-once of the tokens from a finalized channel outcome. `NITRO_MAX_GAS / 20000` would be a sensible choice. Remember to leave some headroom for the `transfer` method's intrinsic gas costs: our test suite confirms that at least 100 Token payouts are possible.
+
+TLDR: stick to outcomes withe fewer than `MAX_OUTCOME_ITEMS` entries, and don't try to `transfer` many more than `NITRO_MAX_GAS` / 20000 tokens in one `transfer` transaction.
```
Comment (snario) [inline packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:126]:
```suggestion
TLDR: stick to outcomes with fewer than `MAX_OUTCOME_ITEMS` entries, and don't try to `transfer` many more than `NITRO_MAX_GAS` / 20000 tokens in one `transfer` transaction.
```

File: packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md [hunk 2]
```diff
@@ -107,9 +107,23 @@ As a state channel participant, it is advised to check the `FixedPart` of any ch
...
+- `channelNonce` -- This should be different to any previous channelNonce used by the same `participants` and `chainId`. This is to prevent states from previous channels being "replayed" to conclude subsequent channels with unintended outcomes. Must be less than `2**48-1`.
 - `appDefinition` -- There should be a [`ForceMoveApp`]('contract-api/natspec/ForceMove') compliant contract deployed at this address, and you should have confidence that it is not malicious or suffering from security flaws. You should inspect the source code (which should be publically available and verifiable) or appeal to a trusted authority to do this.
-- `challengeDuration` -- In the extreme, this should be at least 1 block time (15 seconds on mainnet) and less than `2**48-1` seconds. In practice we recommend somewhere between 5 minutes and 5 months.
+- `challengeDuration` -- In the extreme, this should be at least 1 block time (15 seconds on mainnet) and less than `2**48-1` seconds. Whatever it is set to, the channel should be closed long before `2**48 - 1 - challengeDuration`. In practice we recommend somewhere between 5 minutes and 5 months.
+
+### Limits
```
Comment (snario) [inline packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:114]: It might be worth writing a Github issue in the repo related to working around this limitation. We can do some interesting things in the wallet automatically like triggering a "bundling" of some group of channels after we reach some threshold. For example if a ledger channels funds `MAX_OUTCOME_ITEMS / 2` outcomes, then for the next one maybe it creates a new "sub ledger" channel funding these application channels, and then funds that sub ledger channel.

Just suggesting we keep track of this general possibility; maybe one day it becomes a priority, or someone interested in the project might want to tackle it.

**Comment by andrewgordstewart** [comment]
Does this fix https://github.com/statechannels/project-management/issues/3?

**Comment by geoknee** [comment]
I would like it to be judged against that issue, but I don't think I can link the PR to it because it is in a different repository.

File: packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md (hunk 2 above)
Comment (andrewgordstewart) [reply packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:114]: This exact strategy is in my (draft) roadmap 😉

**Comment by andrewgordstewart** [comment]
I think that putting
> [Andrew] ... let's see if this
> Fixes statechannels/project-management#3
in the PR description lead to this:

Screen Shot 2020-12-04 at 3 49 09 PM

File: packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md [hunk 3]
```diff
@@ -107,9 +107,23 @@ As a state channel participant, it is advised to check the `FixedPart` of any ch
...
+- `channelNonce` -- This should be different to any previous channelNonce used by the same `participants` and `chainId`. This is to prevent states from previous channels being "replayed" to conclude subsequent channels with unintended outcomes. Must be less than `2**48-1`.
 - `appDefinition` -- There should be a [`ForceMoveApp`]('contract-api/natspec/ForceMove') compliant contract deployed at this address, and you should have confidence that it is not malicious or suffering from security flaws. You should inspect the source code (which should be publically available and verifiable) or appeal to a trusted authority to do this.
-- `challengeDuration` -- In the extreme, this should be at least 1 block time (15 seconds on mainnet) and less than `2**48-1` seconds. In practice we recommend somewhere between 5 minutes and 5 months.
+- `challengeDuration` -- In the extreme, this should be at least 1 block time (15 seconds on mainnet) and less than `2**48-1` seconds. Whatever it is set to, the channel should be closed long before `2**48 - 1 - challengeDuration`. In practice we recommend somewhere between 5 minutes and 5 months.
```
Comment (andrewgordstewart) [inline packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:112]: I'm not sure about this specific suggestion, but maybe we can have a suggestion in this vein somewhere [here](https://github.com/statechannels/statechannels/pull/3023/files#diff-68056c6c283f5e38bd767476fe7ce18e3d0bba2c14f540c5ecd9ce1456e1cedeL104-L106)?

```suggestion
- `challengeDuration` -- In the extreme, this should be at least 1 block time (15 seconds on mainnet) and less than `2**48-1` seconds. Whatever it is set to, the channel should be closed long before `2**48 - 1 - challengeDuration`. In practice we recommend somewhere between 5 minutes and 5 months. Consult the @statechannels team for more context.
```

**Review by andrewgordstewart — APPROVED** [review]

File: packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md (hunk 2 above)
Comment (geoknee) [reply packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:114]: Strongly support this: I am minded to not bother with the GH issue if @andrewgordstewart is tracking it elsewhere, though? I feel we suffer a bit from Issue inflation and it might get lost or forgotten about.

Comment (andrewgordstewart) [reply packages/docs-website/docs/protocol-tutorial/execute-state-transitions.md:114]: > I feel we suffer a bit from Issue inflation and it might get lost or forgotten about.

To be a bit pedantic, issue inflation won't cause issues to become lost, but might cause them to be forgotten.

To be a bit more pedantic,
> lost **or** forgotten about

is still true given the first bit of pedantry.
